### Exercise 2.10: The semi-empirical mass formula

In nuclear physics, the semi-empirical mass formula is a formula for calculating the approximate nuclear binding energy $B$ of an atomic nucleus with atomic number $Z$ and mass number $A$:
\begin{gather}
    B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A-2Z)^2}{A} + a_5\frac{1}{A^{1/2}}
\end{gather}
where, in units of millions of electron volts (MeV), the constants are $a_1 = 15.8$, $a_2 = 18.3$, $a_3 = 0.714$,
$a_4 = 23.2$, and
\begin{gather}
    a_5 = 
    \begin{cases}
    0 &  &  & \text{if $A$ is odd}\\
    12 &  &  & \text{if $A$ and $Z$ are both even}\\
    -12 &  &  & \text{if $A$ is even and $Z$ is odd.}
    \end{cases}
\end{gather}
**a)** Write a program that takes as its input the values of $A$ and $Z$, and prints out the binding energy for the corresponding atom. Use your program to find the binding energy of an atom with $A = 58$ and $Z = 28$. (Hint: The correct answer is around $490$ MeV.)

***

In [3]:
# Cell 1

    # User input  
A = float(input('Indicate the mass number A:'))
Z = float(input('Indicate the atomic number Z:'))

    # Defining constants 
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
if A % 2 == 1:
    a5 = 0 
elif A % 2 == 0 and Z % 2 == 0:
    a5 = 12
elif A % 2 == 0 and Z % 2 == 1:
    a5 = -12 

    # Binding energy semi-empirical formula
B = a1*A - a2*A**(2/3) - a3*(Z**2/A**(1/3)) - a4*((A-2*Z)**2/A)+ a5/A**(1/2)

    # Output 
print('The Binding energy for this nucleus is:',f"{B:,.2f}", 'MeV')

Indicate the mass number A: 58
Indicate the atomic number Z: 28


The Binding energy for this nucleus is: 497.56 MeV


**b)** Modify your program to print out not the total binding energy $B$, but the binding energy per nucleon, which is $B/A$.

***

In [4]:
# Cell 2

    # User input  
A = float(input('Indicate the mass number A:'))
Z = float(input('Indicate the atomic number Z:'))

    # Defining constants 
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2
if A % 2 == 1:
    a5 = 0 
elif A % 2 == 0 and Z % 2 == 0:
    a5 = 12
elif A % 2 == 0 and Z % 2 == 1:
    a5 = -12 

    # Binding energy semi-empirical formula
B = a1*A - a2*A**(2/3) - a3*(Z**2/A**(1/3)) - a4*((A-2*Z)**2/A)+ a5/A**(1/2)

    # Output 
print('The Binding energy per nucleon for this nucleus is:',f"{B/A:,.2f}", 'MeV')

Indicate the mass number A: 58
Indicate the atomic number Z: 28


The Binding energy per nucleon for this nucleus is: 8.58 MeV


**c)** Now modify your program so that it takes as input just a single value of the atomic number $Z$ and then goes through all values of $A$ from $A = Z$ to $A = 3Z$, to find the one that has the largest binding energy per nucleon. This is the most stable nucleus with the given atomic number. Have your program print out the value of $A$ for this most stable nucleus and the value of the binding energy per nucleon.

***

In [11]:
# Cell 3

    # User input
Z = float(input('Indicate the atomic number Z:'))
A = Z

    # Defining constants 
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2

    # Defining empty lists 
binding_per_nucleon_list = [0]*(2*int(Z) + 1) # list of binding energy per nucleon 
mass_number_list = [0]*(2*int(Z) + 1) # list of corresponding mass number

i = 0 # index to be used 

    # loop that goes over all nucleus with Z < A < 3Z
while(A <= 3*Z):
    
        # Defining last constant
    if A % 2 == 1:
        a5 = 0 
    elif A % 2 == 0 and Z % 2 == 0:
        a5 = 12
    elif A % 2 == 0 and Z % 2 == 1:
        a5 = -12 
 
        # Binding energy semi-empirical formula
    B = a1*A - a2*A**(2/3) - a3*(Z**2/A**(1/3)) - a4*((A-2*Z)**2/A)+ a5/A**(1/2)

        # Filling lists with information
    binding_per_nucleon_list[i] = B/A 
    mass_number_list[i] = A
    i = i + 1
    A = A + 1

    # output 
print('The most stable nucleus for Z =',Z,',')
print('has a binding energy per nucleon on:',max(binding_per_nucleon_list),'MeV.')
print('This corresponds to mass number of',mass_number_list[binding_per_nucleon_list.index(max(binding_per_nucleon_list))],'.')


Indicate the atomic number Z: 28


The most stable nucleus for Z = 28.0 ,
has a binding energy per nucleon on: 8.70245768367189 MeV.
This corresponds to mass number of 62.0 .


**d)** Modify your program again so that, instead of taking $Z$ as input, it runs through all values of $Z$ from $1$ to $100$ and prints out the most stable value of A for each one. At what value of $Z$ does the maximum binding energy per nucleon occur? (The true answer, in real life, is $Z = 28$, which is nickel. You should find that the semi-empirical mass formula gets the answer roughly right, but not exactly.)

***

In [3]:
#defino las constantes
a = 15.8
b = 18.3
c = 0.714
d = 23.2

L = [0]*100
S = [0]*100

j = 1.00
for Z in [*range(1,101)]:
    A = Z
    l = [0]*(2*Z + 1)
    s = [0]*(2*Z + 1)
    i = 0
    
    while(A <= 3*Z):

        if A % 2 == 1:
            d = 0 
        elif A % 2 == 0 and Z % 2 == 0:
            d = 12
        elif A % 2 == 0 and Z % 2 == 1:
            d = -12  
    
        B = a*A - b*A**(2/3) - c*(Z**2/A**(1/3)) - d*((A-2*Z)**2/A)+ c/A**(1/2)
    
        l[i] = B/A 
        s[i] = A
        i = i + 1
        A = A + 1
        
    L[Z-1] = max(l)
    S[Z-1] = s[l.index(max(l))]
    if(max(l)>j):
        j = max(l)
        z = Z

print('The nucleous Z = ',z,',')
print('has a higher binding energy per nucleon of,',max(L),'MeV.')
print('The A Value that gives us thus value is A =',S[L.index(max(L))],'.')


The nucleous Z =  23 ,
has a higher binding energy per nucleon of, 14.089177168639827 MeV.
The A Value that gives us thus value is A = 24 .


### Exercise 2.6: Planetary orbits

The orbit in space of one body around another, such as a planet around the Sun, need not be circular. In general it takes the form of an ellipse, with the body sometimes closer in andsometimes further out. If you are given the distance $l_1$ of closest approach that a planet makes to the Sun, also called its perihelion, and its linear velocity $v_1$ at perihelion, then any other property of the orbit can be calculated from these two as follows.

**a)** Kepler's second law tells us that the distance $l_2$ and velocity $v_2$ of the planet at its most distant point, or apnetion, satisty $l_1 v_1 = l_2 v_2 $. At the same time the total energy, Kinetic plus gravitational, of a planet with velocity $v$ and distance $r$ from the Sun is given by
\begin{gather}
    E = \frac{1}{2}mv^2 - G\frac{mM}{r}
\end{gather}
where $m$ is the planet's mass, $M = 1.9891 \times 10^{30}$ kg is the mass of the Sun, and $G = 6.6738 × 10^{-11} Nm^2$ is Newton's gravitational constant. Given that energy must be conserved, show that $v_2$ is the smaller root of the quadratic equation
\begin{gather*}
    v_2^2 - \frac{2GM}{v_1l_1}v_2 - \left( v_1^2 - \frac{2GM}{l_1}  \right)=0
\end{gather*}
Once we have $v_2$ we can calculate $l_2$ using the relation $l_2 = l_1 v_1/v_2$.

***

Given the following equation: 
\begin{gather*}
    v_2^2 - \frac{2GM}{v_1l_1}v_2 - \left( v_1^2 - \frac{2GM}{l_1}  \right)=0
\end{gather*}
By using the quadratic formula we obtain:  
\begin{gather}
    v_2 = \frac{(2GM/v_1l_1) \pm (2v_1 - (2GM/v_1l_1))}{2}
\end{gather}
which reduces to 
\begin{align*}
    v_2 & = v_1 & v_2 & = (2GM/v_1l_1) - v_1
\end{align*}
We know that $l_1v_1 = l_2v_2$, and by definition $l_1 < l_2$. as such we must choose the second root

**b)** Given the values of $v_1$, $l_1$, and $l_2$, other parameters of the orbit are given by simple formulas can that be derived from Kepler's laws and the fact that the orbit is an ellipse:
\begin{align}
    \text{Semi-major axis:}& & a&= (l_1 + l_2)/2 \\
    \text{Semi-minor axis:}& & b&= \sqrt{l_1 l_2} \\
    \text{Orbital period:}& & T&= \frac{2\pi ab}{l_1v_1}\\
    \text{Orbital eccentricity:}& & e&= \frac{l_2-l_1}{l_2+l_1}
\end{align}
Write a program that asks the user to enter the distance to the Sun and velocity at perihe-lion, then calculates and prints the quantities $l_2$, $v_2$, $T$, and $e$.

***

In [3]:

from math import pi

    # Constants 
M = 1.9891e30
G = 6.66738e-11

    # User input
l1 = float(input('Cual es el valor de l1?'))
v1 = float(input('Cual es el valor de v1?'))

    # Calculations
v2 = (2*G*M) / (v1*l1) - v1
l2 = (l1*v1) / v2

a = 0.5*(l1+l2) # Semi-major axis (meters)
b = (l1*l2)**(0.5) # Semi-minor axis (meters)
T = (2*pi*a*b)/(l1*v1) # Orbital period (second)
e = (l2-l1)/(l2+l1) # Orbital eccentricity

print('The value of l2=',f"{l2:,.2f}",'m')
print('The value of v2=',f"{v2:,.2f}",'m/s')
print('The Orbital period is T=',f"{T/(31_556_952):,.2f}","years")
print('The orbital eccentricity is e =', f"{e:,.4f}")

Cual es el valor de l1? 1.4710e11
Cual es el valor de v1? 3.0287e4


The value of l2= 152,325,170,090.45 m
The value of v2= 29,248.07 m/s
The Orbital period is T= 1.00 years
The orbital eccentricity is e = 0.0175


**c)** Test your program by having it calculate the properties of the orbits of the Earth (for which $l_1= 1.4710 \times 10^{11} m$ and $v_1 = 3.0287 \times 10^4 m/s$) and Halley's comet ($l_1 = 8.7830 \times 10^{10} m$ and $v_1 = 5.4529 \times 10^4 m/s$). Among other things, you should find that the orbital period of the Earth is one year and that of Halley's comet is about $76$ years. 

***

Running the program for planet Earth we have the following output:  
The value of l2= 152,325,170,090.45 m   
The value of v2= 29,248.07 m/s    
The Orbital period is T= 1.00 years   
The orbital eccentricity is e = 0.0175  

Running the program for Halley's comet we have the following output:   
The value of l2= 5,612,309,285,034.12 m  
The value of v2= 853.35 m/s  
The Orbital period is T= 83.19 years  
The orbital eccentricity is e = 0.9692